### ISSUE TITLE
Support multiple autoload paths for a single namespace

### ISSUE THREAD

**Issue by cretz** [goaop/framework#56]
Composer supports providing an array of autoloading paths for a single namespace. In WeavingTransformer though, array_map is called for "realpath". That assumes all values are strings. It should be "array_map_recursive" type thing that at least goes one more array deep.

**Comment by lisachenko** [comment]
@cretz I think that for the current version of Go! AOP library it will be better to configure `autoloadPaths` by hand.

I know about the problem with composer and have some ideas about hijacking it ) But this work in progress. 
Look at https://github.com/lisachenko/go-aop-php/commit/cf7087e98fe3b3d2aa6ee0efc5b7f9269de52ac9 if you want some ideas about how it will work with composer.

**Comment by cretz** [comment]
My composer autoloader initalizer is even more complex than that. I do configure it by hand. My problem is that a namespace prefix can have MULTIPLE paths yet the WeavingTransformer assumes the values for the namespace array are always just strings. For instance, I should be able to have this:

```
<?php
$kernel->init([
    'autoloadPaths' => ['My\Namespace' => ['my-project/src', 'my-project/tests']]
]);
```

This is akin to the monolog example on the composer site: http://getcomposer.org/doc/04-schema.md#autoload. I hope that makes sense.

**Comment by lisachenko** [comment]
Thank you for feedback. Probably I found the issue with your example. You shouldn't use autoload paths in the AOP kernel to load custom classes. This section is used only for internal classes and namespaces: Go, Dissect, Doctrine and TokenReflection.
Separate configuration for internal classes is needed to prevent framework from weaving aspects into internal classes. So, it shouldn't be used for external classes.

**Comment by cretz** [comment]
That's exactly how I ran into it, I want to avoid other unimportant packages/code when intercepting my own code. I use https://packagist.org/packages/phpdocumentor/reflection which requires https://packagist.org/packages/phpdocumentor/reflection-docblock and I do not want any of these to be subject to any AOP. Note, https://github.com/phpDocumentor/ReflectionDocBlock/blob/master/composer.json#L13 and https://github.com/phpDocumentor/Reflection/blob/master/composer.json#L9 use the same namespace on different paths. I have a hack in place to fix these in my own code so the autoloadPaths array is always string => string instead of the possibility of string => string|string[]:

```
/** @return string[] */
private function getHackedExcludedPaths()
{
    // TODO: Sadly, https://github.com/lisachenko/go-aop-php/issues/56 prevents
    //  the autoload paths from having nested arrays, so we hack it here for
    //  to get a common path just after vendor/
    $properExcludedPaths = [];
    foreach ($this->excludedPaths as $namespace => $excludedPath) {
        if (is_array($excludedPath)) {
            echo("WARN: {$namespace} has multiple paths, using one base one\n");
            $path = $excludedPath[0];
            $endIndex = strpos($path, 'vendor/');
            if ($endIndex === false) {
                throw new \Exception('Unable to find vendor/');
            }
            $endIndex = strpos($path, '/', $endIndex + 7);
            if ($endIndex === false) {
                throw new \Exception('Unable to find dir after vendor');
            }
            $properExcludedPaths[$namespace] = substr($path, 0, $endIndex);
        } else {
            $properExcludedPaths[$namespace] = $excludedPath;
        }
    }
    return $properExcludedPaths;
}
```

Basically, I believe there is a need for a single namespace to have multiple paths. Am I doing something wrong when excluding these paths?

**Comment by lisachenko** [comment]
Ok, If you want an AOP only for your own code, then it's better to specify a whitelist of directories where AOP will be applied. Look at the `includePaths` option, it defines folders where AOP should be applied. 
If it's empty, then each file under the application directory will be processed.

Hope, that this option will help you to solve this issue )

**Comment by lisachenko** [comment]
P.S. Can I suggest you to use [andrewsville/php-token-reflection](https://packagist.org/packages/andrewsville/php-token-reflection) for static reflection? My framework works with it, so if you choose it too, then overall performance will be better and project size will be smaller.

**Comment by cretz** [comment]
I will use includePaths, though I was wanting to make this a tool that composer users could plugin and I could just pull autoload namespaces, but that requires properly parsing multiple paths for a single namespace. Another thing I was kinda hoping for was the ability to add include/autoload paths AFTER init. My main goal was to be a drop-in replacement for composer so people's autoloading scripts could be like:

```
$loader = require(isset($_ENV['contract'] ? 'vendor/myproj/autoload.php' : 'vendor/autoload.php');
$loader->add('MyNs', __DIR__);
```

I very well might switch over to that reflection lib. My main two needs are docblock tag parsing (not the contents of each tag, just the tag itself) and the ability to preserve namespace aliases so I can resolve unqualified class names in phpdoc. I see that lib has both and is so much lighter weight than the phpdoc one, so that makes me happy.

So, just to confirm, the autoload paths will not be supporting multiple paths for a single name? Maybe I'll submit a pull request if I get the chance.

**Comment by lisachenko** [comment]
I want to make better integration with composer too, and I tried to make some patches for this, but there are some caveats that prevent me from merging commits into the main branch. However, commit https://github.com/lisachenko/go-aop-php/commit/9b7982f706779a5b5669505de44cec5a99218a83 introduces the compatibility with composer.

My plans is to remove `autoloadPath` configuration option and use it only for cases when composer is not available in the application. If composer is available, then framework will reuse it with custom logic.
